rfd/0003-workspace-ram-management.md

```diff
@@ -0,0 +1,70 @@
+---
+authors: Grzegorz Uriasz (gorbak25@gmail.com)
+state: draft
+---
+
+# RFD 3 - Workspace RAM Management
+
+## What
+
+Specify RAM management in workspaces so Hocus VMs feel like containers.
+
+## Why
+
+To increase the density of workspaces on a single machine and to minimize the footprint of workspaces.
+
+## Details
+
+### Containers vs MicroVM's vs Hocus Workspaces
+
+Containers are much more lightweight than VM's due to their shared host kernel and page cache. When running a container, the workload has full access to the host resources, and efficient resource sharing is easy. VMs, on the other hand, run a fully separate kernel and are always designed to use every MB of RAM that was given to them. Workloads running in VMs, contrary to containers, might be using up 1GB of RAM but 15GB in various caches, which are never freed and usually duplicated on the host kernel, resulting in very high memory usage on the host and decreasing the density of workspaces. MicroVM's are short-lived, so they avoid the problem of page caches piling up altogether. Hocus workspaces are meant to be operating for a very long time and have periods of inactivity, like a developer having 3 workspaces open but only using 1 of them or leaving workspaces on during their lunch break or overnight. Hocus workspaces are designed to autoscale their RAM during periods of intense usage (like compiling a project) and to release back their RAM as soon as the usage comes down.
+
+### Manual ballooning
+
+Previously, the only option for dynamic VM RAM management was memory ballooning where the hypervisor might manually readjust the size of the VM's RAM. Unfortunately, this introduces a ton of complexity in the hypervisor and is a half measure, not a solution, as creating a RAM usage policy which actually works well is very hard. Even if a good policy was designed and implemented, the inflation/deflation of the balloon is slow, unreliable, and unsuited for our use case. We want VMs to access most of the Host resources like containers and quickly give them back when unneeded.
+
+### Free page reporting
+
+The solution to this problem is Free Page Reporting[^1], a feature in `virtio-balloon` available since the 5.7 kernel, which gives a primitive in which the VM might give back their unused RAM to the hypervisor without manual intervention. By default, free page reporting operates in 4MB segments on x86_64, limiting its effectiveness without other measures. Since the 5.14 kernel, it's possible to override the segment size and for free page reporting to operate on a per-page granularity[^2]. In my testing, free page reporting with `page_reporting.page_reporting_order=0` works well and has acceptable overhead.
```

Comment:
```suggestion
The solution to this problem is Free Page Reporting[^1], a feature in `virtio-balloon` available since the 5.7 kernel, which provides a primitive in which the VM can give back its unused RAM to the hypervisor without manual intervention. By default, free page reporting operates in 4MB segments on x86_64, limiting its effectiveness without other measures. Since the 5.14 kernel, it's possible to override the segment size and for free page reporting to operate on a per-page granularity[^2]. In my testing, free page reporting with `page_reporting.page_reporting_order=0` works well and has acceptable overhead.
```